### ISSUE TITLE
The library 'hostpolicy.dll' required to execute the application was not found in

### ISSUE THREAD

**Issue by peterhuene** [dotnet/core-setup#6189]
_From @sillyousu on Tuesday, 30 April 2019 02:19:35_

## Steps to reproduce

create a hello world project:
```
mkdir ClassLib
cd ClassLib
dotnet new classlib
cd ..
mkdir Main 
cd Main
dotnet new console
dotnet  add reference ..\ClassLib\ClassLib.csproj
dotnet build
dotnet .\bin\Debug\netcoreapp2.2\ClassLib.dll
```
```
.
├── ClassLib
│   ├── Class1.cs
│   └── ClassLib.csproj
└── Main
    ├── Main.csproj
    └── Program.cs
```

## Expected  behavior
meaningful error message

## Actual behavior
```
A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'C:\Users\supei\xd\sims\fooo\Main\bin\Debug\netcoreapp2.2\'.
Failed to run as a self-contained app. If this should be a framework-dependent app, add the C:\Users\supei\xd\sims\fooo\Main\bin\Debug\netcoreapp2.2\FooA.runtimeconfig.json file specifying the appropriate framework.
```

## Environment data
`dotnet --info` output:
```
PS C:\Users\supei\xd\sims\fooo\Main> dotnet.exe --info
.NET Core SDK（反映任何 global.json）:
 Version:   2.2.100
 Commit:    b9f2fa0ca8

运行时环境:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.100\

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

.NET Core SDKs installed:
  1.1.8 [C:\Program Files\dotnet\sdk]
  1.1.9 [C:\Program Files\dotnet\sdk]
  1.1.11 [C:\Program Files\dotnet\sdk]
  2.1.101 [C:\Program Files\dotnet\sdk]
  2.1.103 [C:\Program Files\dotnet\sdk]
  2.1.104 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.2.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

_Copied from original issue: dotnet/cli#11267_

**Comment by peterhuene** [comment]
@sillyousu thank you for reporting this issue.

@vitek-karas is there something we could do in the host to have a more helpful error message for users when someone accidentally execs a `netstandard` assembly?

**Comment by borysl** [comment]
I've reproduced this issue when was running xunit tests with the project reference

**Comment by WeihanLi** [comment]
I met the error with dot net framework project

**Comment by vitek-karas** [comment]
@borysl @WeihanLi - I'm just trying to understand your comments better:
* Is it that you've ran into this by also trying to run a non-app project and are simply reporting the fact that the error message is confusing?
* Is it that you were trying to do something you think should work and it failed?

The scenario described above is supposed to fail, this issues is about the fact that the failure message is rather confusing.

**Comment by WeihanLi** [comment]
Thanks for your reply @vitek-karas , this is not exactly the same with my scene, I'm running a web project with dotnet framework, so I opened a new issue here <https://github.com/dotnet/core-setup/issues/7149>

**Comment by AbhishekHumagain** [comment]
Add this json to the **application.runtimeconfig.json** file.
{
  "runtimeOptions": {
    "tfm": "netcoreapp2.2",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "2.0.0"
    }
  }
}

**Comment by WeihanLi** [comment]
I don't think that's bug or something need change, you can not run `dotnet`/`dotnet exec` with a class library, the class library had no entry point for a executable program

**Comment by joanna1010** [comment]
I am having the same issue when deploying my Function app in Azure.

Going into https://my-web-app.scm.azurewebsites.net
and trying to do dotnet myapp.dll
here's the output

![image](https://user-images.githubusercontent.com/5986383/63736593-b8796c80-c8d7-11e9-9dd8-ab3c5b16fe75.png)

By default, Function app boilerplate makes it as class library and you can deploy that dll straight into Azure, means it must be able to be executed properly right?

**Comment by Smurf-IV** [comment]
> 
> 
> I've reproduced this issue when was running xunit tests with the project reference

Same here via a `azure-pipelines.yml` file. Needing to run VSTest@2 due to needing to use rosyln against a .netcore 2.2 test dll that also performs runtime compile tests against output files.
Here is the yml extract:
```yml
- task: VSTest@2
  inputs:
    testSelector: 'testAssemblies'
    testAssemblyVer2: |
             **\*test*.dll
             !**\*TestAdapter.dll
             !**\obj\**
    runInParallel: true
    codeCoverageEnabled: true
    publishRunAttachments: true
    configuration: $(buildConfiguration)
#    runSettingsFile: 'Test.runsettings'
    testRunTitle: $(Agent.JobName)
    timeoutInMinutes: 20
    diagnosticsEnabled: true
    collectDumpOn: 'always'
```

I then get the following for each of the Test.dlls
```

##[error]Testhost process exited with error: A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'd:\a\1\s\Tests\CSharp.Tests\bin\Release\netcoreapp2.2\'.
##[error]Failed to run as a self-contained app. If this should be a framework-dependent app, add the d:\a\1\s\Tests\CSharp.Tests\bin\Release\netcoreapp2.2\testhost.runtimeconfig.json file specifying the appropriate framework.
##[error]. Please check the diagnostic logs for more information.
```
But the files _still_ run against `Xunit`.
**BUT**, because of the errors, the `VSTest` action fails!

**Comment by rob83ert** [comment]
> > I've reproduced this issue when was running xunit tests with the project reference
> 
> Same here via a `azure-pipelines.yml` file. Needing to run VSTest@2 due to needing to use rosyln against a .netcore 2.2 test dll that also performs runtime compile tests against output files.
> Here is the yml extract:
> 
> ```yaml
> - task: VSTest@2
>   inputs:
>     testSelector: 'testAssemblies'
>     testAssemblyVer2: |
>              **\*test*.dll
>              !**\*TestAdapter.dll
>              !**\obj\**
>     runInParallel: true
>     codeCoverageEnabled: true
>     publishRunAttachments: true
>     configuration: $(buildConfiguration)
> #    runSettingsFile: 'Test.runsettings'
>     testRunTitle: $(Agent.JobName)
>     timeoutInMinutes: 20
>     diagnosticsEnabled: true
>     collectDumpOn: 'always'
> ```
> 
> I then get the following for each of the Test.dlls
> 
> ```
> 
> ##[error]Testhost process exited with error: A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'd:\a\1\s\Tests\CSharp.Tests\bin\Release\netcoreapp2.2\'.
> ##[error]Failed to run as a self-contained app. If this should be a framework-dependent app, add the d:\a\1\s\Tests\CSharp.Tests\bin\Release\netcoreapp2.2\testhost.runtimeconfig.json file specifying the appropriate framework.
> ##[error]. Please check the diagnostic logs for more information.
> ```
> 
> But the files _still_ run against `Xunit`. (See the end)
> **BUT**, because of the errors, the `VSTest` action fails!
> 
> ```
> [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (64-bit .NET Core 4.6.28008.02)
> [xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (64-bit .NET Core 4.6.28008.02)
> [xUnit.net 00:00:06.55]   Discovering: 2CSharp.Tests
> [xUnit.net 00:00:06.64]   Discovered:  CSharp.Tests
> [xUnit.net 00:00:06.65]   Starting:    CSharp.Tests
> [xUnit.net 00:00:07.27]   Finished:    CSharp.Tests
> 
> 
>   √ LaTeX2CSharp.Tests.CSharpTests.A100_SimplePower [1ms]
>   √ LaTeX2CSharp.Tests.CSharpTests.A050_SimpleRoot [1ms]
>   ! LaTeX2CSharp.Tests.CSharpTests.A140_NumberInVarName [1ms]
> etc.
> ```

I have the exact same problem...

**Comment by Smurf-IV** [comment]
I found the solution to "My" problem.
Notice the following line contains `**\*test*.dll`
This means that "Any" file that has _test_ in the filename and is a dll will be used.

XUnit places a file called `testhost.dll` in the output directory, which matches the glob and is based on `.netcore 1.x` hence the need of the runner to attempt to use that as the runner policy for that _test.dll_

I changed the filer to be `**\*test.dll` (removed the last `*`) and now all is happy again; i.e. the first message is not displayed and th xunit runner still works as expected.

**Comment by Interface007** [comment]
I came to this page because of having the same message in my logs.

TLDR: 
I was able to solve my problem by removing a project reference from one test-assembly to another. 

Long description with a hint about searching for causes of failed builds:
Have the same issue with an in azure hosted build agent and with an on-prem build agent when running a VSTest task:
```
- task: VSTest@2
  inputs:
    testSelector: 'testAssemblies'
    testAssemblyVer2: |
        **\YYY.*.Tests.dll
        !**\obj\**
    searchFolder: '$(System.DefaultWorkingDirectory)'
    diagnosticsEnabled: true
    codeCoverageEnabled: true
```
As you can see in the task definition, it specifies to only get files matching "YYY.*.Tests.dll", which only includes my test projects.

Unfortunately, when clicking on the link of the commit hash of the "run", you see the changes of that particular commit - not the changes between ins run and the previous run. Between the last "successful run" and the first "failing run" I did several commits locally that I pushed to Azure DevOps in one single push operation.

So, the changes shown with the "failing run" may not include the cause of the failure - you have to find the changes between the two commits manually.

**Comment by MichaelSchreiber** [comment]
As mentioned in the preceding posts, in most cases the actual source of the issue seems to be that **\*test*.dll is the default when creating a VSTest task. This includes testhost.dll and the test task tries to execute it as a test dll.
One way to reduce the frequency in wich this issue occurs is to add the exclusion !**\*testhost.dll to the default for the VSTest task as it is already done with !**\*TestAdapter.dll.
Concerning a better error message, adding something like "If this error occurs in a test task, excluding testhost.dll via !**\*testhost.dll may solve the issue" might help users to find a solution pretty fast.

**Comment by GitSnafu** [comment]
I just got hit by this issue after upgrading my app to netcoreapp3.1 from netcoreapp3.0.

In my case the hostpolicy.dll simply doesn't exist in C:\Program Files\dotnet\shared\Microsoft.NETCore.App\3.1.0. In fact there's only a single file in that folder so it looks like the 3.1 runtime installation is incomplete/corrupted. Note that it still comes up as being installed when running `dotnet --list-runtimes`.

```
Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All] 
Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App] 
Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App] 
Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App] 
Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App] 
Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App] 
```

**Comment by vitek-karas** [comment]
@GitSnafu Would be interesting to know how you installed it (and if the installation failed). The way dotnet checks for frameworks is simply by looking for a folder with a given name, it doesn't check the contents (it trusts the installer to do its job). So it's not surprising that it reports the framework as available even if there's nothing in it.

What file is in that folder?

You should be able to fix this by reinstalling 3.1 I think.

**Comment by GitSnafu** [comment]
@vitek-karas Yes, re-installing it most likely will fix the issue, which is why I thought it was worth mentioning it in case anyone else encounters this error for similar reasons.

The only file in that folder was mscordaccore_amd64_amd64_4.700.19.56402.dll.

I believe it was installed using Ansible. All other servers I've checked that would've also had it installed in the same way look good, so it seems like there was just an issue with this particular instance.

**Comment by vitek-karas** [comment]
So far I was only able to determine that such file is used by dump debugging (!analyze in windbg). Would be interesting to know how exactly it got installed, as such action is effectively breaking for the .NET Core installation on the machine.

**Comment by AudricCh** [comment]
Hi,

I wanna know where we are about this.

I'm trying to use a dotnet command to generate a swagger file using SwashBuckle
> dotnet swagger tofile --output ./ ./MyApi/bin/Debug/net47/MyApi.dll swagger.json

And I've got the following message
> A fatal error was encountered. The library 'hostpolicy.dll' required to execute the application was not found in 'D:\MyApi\bin\Debug\net47\'.
> Failed to run as a self-contained app. If this should be a framework-dependent app, specify the appropriate framework in D:\MyApp\bin\Debug\net47\MyApi.runtimeconfig.json.

And the MyApi.runtimeconfig.json file looks great :
`{
  "runtimeOptions": {
    "tfm": "net47",
    "configProperties": {
      "System.GC.Server": true
    }
  }
}`

Is it just about the NetCore / NetStandard compatibility ?
Thanks

**Comment by vitek-karas** [comment]
@AudricCh You're trying to run .NET Framework application on .NET Core - that is not going to work. The `net47` TFM is .NET Framework 4.7.

The interesting question in this case would be how did you build the app such that it produced `.runtimeconfig.json` even for `net47` TFM. That should not happen.

Can you please file a new issue for that in https://github.com/dotnet/sdk repo? Thanks a lot!
